modules/zeromq/src/main/java/org/apache/ignite/stream/zeromq/IgniteZeroMqStreamer.java

```diff
@@ -0,0 +1,133 @@
+ * limitations under the License.
+ */
+
+package org.apache.ignite.stream.zeromq;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.util.typedef.internal.A;
+import org.apache.ignite.stream.StreamAdapter;
+import org.jetbrains.annotations.NotNull;
+import org.zeromq.ZMQ;
+
+/**
+ * This streamer uses https://github.com/zeromq/jeromq/.
+ * Implements socket types listed {@link ZeroMqTypeSocket}.
+ */
+public class IgniteZeroMqStreamer<K, V> extends StreamAdapter<byte[], K, V> implements AutoCloseable {
+    /** Logger. */
+    protected IgniteLogger log;
+
+    /** Parameter {@code True} if streamer started. */
+    private volatile boolean isStarted = false;
+
+    /** Process stream asynchronously */
+    private ExecutorService zeroMqExSrv;
+
+    /** ZeroMQ context. */
+    private ZMQ.Context ctx;
+
+    /** ZeroMQ socket. */
+    private ZMQ.Socket socket;
+
+    /** ZeroMQ context threads. */
+    private int ioThreads;
+
+    /** ZeroMQ socket type. */
+    private int socketType;
+
+    /** ZeroMQ socket address.*/
+    private String addr;
+
+    /** ZeroMQ topic name. */
+    private byte[] topic;
+
+    /** Maximum time to wait (time unit minutes). */
+    private int timeout = 2;
+
+    /**
+     * @param ioThreads Threads on context.
+     * @param socketType Socket type.
+     * @param addr Address to connect zmq.
+     * @param topic Topic name for PUB-SUB socket type, otherwise null.
+     */
+    public IgniteZeroMqStreamer(int ioThreads, ZeroMqTypeSocket socketType, @NotNull String addr, byte[] topic) {
+        A.ensure(ioThreads > 0, "Param ioThreads has been more than 0.");
+        A.ensure(!"".equals(addr), "Param addr has been not empty.");
+        A.ensure(socketType != null, "This socket type is not implemented in this version of ZeroMQ streamer.");
+
+        this.ioThreads = ioThreads;
+        this.addr = addr;
+        this.topic = topic;
+        this.socketType = socketType.getType();
+    }
+
+    /**
+     * Start ZeroMQ streamer.
+     */
+    public void start() {
+        A.ensure(getSingleTupleExtractor() != null || getMultipleTupleExtractor() != null, "ZeroMq extractor.");
+
+        log = getIgnite().log();
+
+        if (isStarted) {
+            log.warning("Attempted to start an already started ZeroMQ streamer");
+            return;
+        }
+
+        isStarted = true;
+
+        ctx = ZMQ.context(ioThreads);
+        socket = ctx.socket(socketType);
+        socket.connect(addr);
+
+        if (ZeroMqTypeSocket.SUB.getType() == socketType)
+            socket.subscribe(topic);
+
+        zeroMqExSrv = Executors.newSingleThreadExecutor();
+
+        Callable<Boolean> task = new Callable<Boolean>() {
+            @Override
+            public Boolean call() {
+                while (true) {
```

Comment:
Getting back to this loop :)
Its completion can be cleaner if you use, for instance, `Runnable` with `isStarted`.
In the current implementation, it ends with `org.zeromq.ZMQException` because ZMQ context is closed. After a quick look at the ZMQ code, I think there are chances the loop is not terminated, if ZMQ context fails to properly terminate.